src/Nethermind/Nethermind.Network/P2P/SyncPeerProtocolHandlerBase.cs

```diff
@@ -0,0 +1,463 @@
+            if (Logger.IsTrace) Logger.Trace($"OUT {Counter:D5} BlockHeaders to {Node:c} in {stopwatch.Elapsed.TotalMilliseconds}ms");
+        }
+
+        protected void Handle(BlockHeadersMessage message, long size)
+        {
+            Request<GetBlockHeadersMessage, BlockHeader[]> request = _headersRequests.Take();
+            if (message.PacketType == Eth62MessageCode.BlockHeaders)
+            {
+                request.ResponseSize = size;
+                request.CompletionSource.SetResult(message.BlockHeaders);
+            }
+        }
+
+        protected void Handle(GetBlockBodiesMessage request)
+        {
+            if (request.BlockHashes.Count > 512)
+            {
+                throw new EthSynchronizationException("Incoming bodies request for more than 512 bodies");
+            }
+
+            if (Logger.IsTrace)
+            {
+                Logger.Trace($"Received bodies request of length {request.BlockHashes.Count} from {Session.Node:c}:");
+            }
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            IList<Keccak> hashes = request.BlockHashes;
+            Block[] blocks = new Block[hashes.Count];
+
+            for (int i = 0; i < hashes.Count; i++)
+            {
+                blocks[i] = SyncServer.Find(hashes[i]);
+            }
+
+            Interlocked.Increment(ref Counter);
+            Send(new BlockBodiesMessage(blocks));
+            stopwatch.Stop();
+            if (Logger.IsTrace) Logger.Trace($"OUT {Counter:D5} BlockBodies to {Node:c} in {stopwatch.Elapsed.TotalMilliseconds}ms");
+        }
+
+        protected void Handle(BlockBodiesMessage message, long size)
+        {
+            Request<GetBlockBodiesMessage, BlockBody[]> request = _bodiesRequests.Take();
+            if (message.PacketType == Eth62MessageCode.BlockBodies)
+            {
+                request.ResponseSize = size;
+                request.CompletionSource.SetResult(message.Bodies);
+            }
+        }
+
+        private static BlockHeader[] FixHeadersForGeth(BlockHeader[] headers)
+        {
+            int emptyBlocksAtTheEnd = 0;
+            for (int i = 0; i < headers.Length; i++)
+            {
+                if (headers[headers.Length - 1 - i] == null)
+                {
+                    emptyBlocksAtTheEnd++;
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            if (emptyBlocksAtTheEnd != 0)
+            {
+                BlockHeader[] gethFriendlyHeaders = headers.AsSpan(0, headers.Length - emptyBlocksAtTheEnd).ToArray();
+                headers = gethFriendlyHeaders;
+            }
+
+            return headers;
+        }
+
+        #region Cleanup
+        private bool _isDisposed;
+        protected abstract void OnDisposed();
+
+        // todo - why can't this just be Dispose()?
+        public override void InitiateDisconnect(DisconnectReason disconnectReason, string details)
```

Comment:
part of the API - so the user classes usually will think about disconnecting as a business action rather disposing -as a tech / cleanup action

Reply:
also there is a delay between initiate disconnect - devp2p requires you to gice some time to other nodes to handle your disconnect message

Reply:
This comment should probably read "why can't this just call Dispose()", and was really just a marker for me to remember to look into it some more - but to explain my thinking at the time..... right now, when `InitiateDisconnect` is called, for instance, `_headersRequests.IsAddingCompleted` is set twice. Once in `InitiateDisconnect`, and once in `Dispose`. If this were just a call to `Dispose` and `Session.InitiateDisconnect`, it would take advantage of the existing logic to prevent disposing twice.

I haven't taken the time yet to look too much into what's happening at the session level, but as far as I can tell, with the exception of `GetBlockHeaders` and `GetBlockBodies` requests, we're still handling incoming transactions from the peer until the disconnect timeout runs out. I guess the protocol isn't super clear about this (using https://github.com/ethereum/devp2p/blob/master/rlpx.md as the source), but it makes sense to me that it would at least behave consistently across message types. I'm not at all familiar with Go, but if I'm understanding this correctly, geth doesn't respond to any sub-protocol messages after sending a disconnect message (https://github.com/ethereum/go-ethereum/blob/4fabd9cbd2b8fc6bdc643fefb5ea583c5019171c/p2p/peer.go#L311)

So, if `InitiateDisconnect` were to call `Disposed` instead of duplicating the `IsAddingCompleted` part, `!_isDisposed` would always signal whether we should bother handling the message.

And at least in the case of Eth62, it seems the only difference between `Dispose`, and `InitiateDisconnect` is that `Dispose` stops the txFlooding timer. This wouldn't be an issue if we're not handling messages from a peer after sending a disconnect message.

Reply:
Agreed, it could be done better, I can see the issues that you are describing here and they may cause trouble.

I am totally fine for you not to solve these issues as they are beyond LES and are part of the technical debt in the networking layer.

I think at the end we can do the review together and collect issues like this one to start new GitHub issues for them.